### DISCUSSION TITLE
getShopReceipt - invalid value for "status"

### DISCUSSION THREAD

**Discussion by studeba** [etsy/open-api#805] [Q&A] [answered]
Hi All... 

First time trying to use the Etsy API.  Using the Swagger (OpenAPI) Ruby client.  I'm able to make several calls without issue but ran into a stumbling block calling getShopReceipt

I'm receiving this response:
invalid value for "status", must be one of ["paid", "completed", "open", "payment processing", "canceled"].

when calling:
entries = api_instance.get_shop_receipt(shop_id, receipt_id)

The docs say only shop_id and receipt_id are required.
https://developers.etsy.com/documentation/reference#operation/getShopReceipt

The generated docs (from Swagger) do not show opts as a param:
https://github.com/studeba/etsy_api/blob/master/docs/ShopReceiptApi.md#get_shop_receipt

Any of ideas of what I'm doing wrong?

Thanks
Mark

**Comment by studeba** [comment]
Update:

We used the OpenAPI at https://www.etsy.com/openapi/generated/oas/3.0.0.json to definition to generate a Ruby client and there is a status property.

 ```
"ShopReceipt": {
      "type": "object",
      "description": "The record of a purchase from a shop. Shop receipts display monetary values using the shop's currency.",
      "properties": {
        ....
       "status": {
              "type": "string",
              "description": "The current order status string. One of: `paid`, `completed`, `open`, `payment processing` or `canceled`.",
              "enum": [
                  "paid",
                  "completed",
                  "open",
                  "payment processing",
                  "canceled"
              ]
          },
```

I've tried generate an API client using both Swagger and OpenAPI - while a little different, they both generate code that validate the Status property

```
    def status=(status)
      validator = EnumAttributeValidator.new('Object', ['paid', 'completed', 'open', 'payment processing', 'canceled'])
      unless validator.valid?(status)
        fail ArgumentError, "invalid value for \"status\", must be one of #{validator.allowable_values}."
      end
      @status = status
    end
```

Could there be something wrong with the API definition?  

Thank

**Reply by Gareth064** [reply]
In the past I have tried to use a code generator to produce an Etsy API Client from the OAS file, I always found too many nuances from the generated code that was causing issues. For example, all ID's from Etsy objects are long integers, but the OAS file defines them as normal integers, so the codes generated would fail because long integers were returned and didn't fit in a normal integer.

This could possible be the another issue from it.

Try commenting out that generated validator (and any reference of it) to see if it makes a difference. 

Its hard to know if the error is being returned from Etsy or from your generated code after the response has already came back from Etsy.

**Comment by studeba** [comment, marked as answer]
Thanks - @Gareth064

The status was coming down proper case.  I modified the generated client to .downcase the status when it's being validated and it seems to be working correctly now.

```
 def status=(status)
    validator = EnumAttributeValidator.new('String', ["paid", "completed", "open", "payment processing", "canceled"])
    unless validator.valid?(status.downcase)
      fail ArgumentError, "invalid value for \"status\", must be one of #{validator.allowable_values}."
    end
    @status = status
  end
```

**Comment by stevenbrookes** [comment]
Hi

We are seeing the same issue.  The OpenAPI file has the status in lowercase and the API returns them in Sentence case.

open-> Open

Does anybody from Etsy look at this repo and update the spec or API?

Regards

Steve

**Reply by Gareth064** [reply]
Unfortunately @stevenbrookes, inconsistencies like this have been raised before back when the Dev community was helping Etsy test the new API, but nothing ever changed.

Best you can do is to email them developer@etsy.com

Or ToLower() everything 🤣

**Comment by stevenbrookes** [comment]
@Gareth064 Thanks for confirming my suspicions.  toLower it is!
